### PR TITLE
Release next

### PR DESCRIPTION
**Pull request by kentcdodds**
**What**: Release the next breaking change

**Why**: Various reasons, but we want people to get this!

**How**: See individual PRs

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged 
- [ ] Added myself to contributors table N/A 

**Todo**:

- [x] merge a few PRs to glamorous
- [x] write blogpost
- [x] [update algolia](https://github.com/algolia/docsearch-configs/blob/master/configs/glamorous.json)
- [x] publish glamorous website
- [ ] merge this PR and wait for the autopublish
- [ ] update the release notes to be more complete and nice
- [ ] publish blogpost

### CODE REVIEW

**Comment by kentcdodds** [comment]
```
npm install glamorous@4.0.0-beta.0
```

💄 v4 🔜

Please try it and submit feedback!

**Comment by kentcdodds** [comment]
Sorry, I squashed all the commits into a single commit. It was much easier to do this than try to manage conflicts when rebasing master. We'll be merging this soon hopefully!

**Comment by kentcdodds** [comment]
I'm going to go ahead and use a comment in here to draft the release notes... @luke-john, could you update this comment (you should have permissions to do that) to include notes about typescript changes?

---

# glamorous@4.0.0 released 💄

## New Features

- `propsAreCssOverrides` is now officially exposed and supported allowing you to create your own version of built-in components (like `glamorous.Div`) that accept props as css. By @kentcdodds in #258
- `withProps` API now available to allow you to specify props you want applied to your component. Give this a look! By @kentcdodds in #255 

## Performance

- @tkh44 found a great optimization in #254 by deferring styles resolution and merging to glamor. Rather than using internal APIs. Benchmarks are silly so it's hard to tell how this will impact your app, but you could see ~8% or more perf improvements when mounting and a bit of a bump for re-rendering.

## Bug Fixes

- In some cases, you could wind up with two glamor-generated classNames on a single glamorous component. Fixed by @tkh44 in #254.

## Breaking Changes

- `glamorous.config` is no longer available. The only configurable option was `useDisplayNameInClassName` and now, if you're using `glamor@2.20.31` or greater then you'll benefit from the new readable classnames feature automatically when `NODE_ENV=development` or `NODE_ENV` is unset. **To update your code** remove any usage of `glamorous.config`. By @oalbacha and @kentcdodds in #221 and #247
- The `theme` argument is no longer available in dynamic props. Instead, `theme` is available on `props`. (Before: `glamorous.div((props, theme) => ({fontSize: theme.fontSize}))` After: `glamorous.div((props) => ({fontSize: props.theme.fontSize}))`. **To update your code** run [this codemod](https://github.com/paypal/glamorous/tree/master/other/codemods#theme-move). By @tdeschryver in #222. Codemod by @tdeschryver and @kentcdodds in #234
- TypeScript: There are tons of TypeScript improvements and updates. Upgrade glamorous and fix all your violations :smile: (:tada: for static typing)



---

Am I missing anything?

**Comment by luke-john** [comment]
the notes appear to be missing mention of shouldClassNameUpdate.

https://github.com/paypal/glamorous/pull/250

**Comment by kentcdodds** [comment]
Good catch, thanks!

**Comment by kentcdodds** [comment]
Anyone wanting to review the blog post, please see the draft here: https://medium.com/@kentcdodds/announcing-the-new-glamorous-website-and-v4-release-c678fe02a39a

Let me know if you have suggestions/improvements/additions/removals/etc :)

**Comment by paulmolluzzo** [comment]
Some notes on the blog post:

Lots of the emoji are not working.
screen shot 2017-07-28 at 9 44 26 pm
screen shot 2017-07-28 at 9 44 30 pm
screen shot 2017-07-28 at 9 44 34 pm

Do you want to wrap references to `glamorous` in code tags? Not sure if that'll become obnoxious after a while. 😅 

> TL;DR: glamorous v4 has been released (glamorous v4 release notes), 

Don't forget the link. 😉 

> There have over 50 code contributors on the core GitHub project and 32 code contributors on the official website repo.

Either "There have been..." or "There are...". I think present tense makes more sense than past perfect here.

> Ther are even more 

Small typo.

> special shoutout to Luke John and Paul M

🙌 . I don't know if you abbreviated my name for some reason, but you're welcome to use my whole name if you like. 😄 

> These things are important to me and I hope this to continue.

Not to put words in your mouth, but I think this could be clearer as: "A welcoming and inclusive open source programming community are important to me, and I hope glamorous continues down this path."

> I’ve felt this is really important to help people learn to use and contribute to the project.

Again, I don't want to put words in your mouth, but I think this reads better as: "I feel that having the documentation in many languages is critical to helping as many people as possible learn to use and contribute to the project."

> Based on those examples, there are three places

"Based on _these_ examples..."

> inspired by jsx-style

A link to that lib would be nice to add (https://github.com/smyte/jsxstyle)

Overall, this is looking great! Super exciting!

**Comment by kentcdodds** [comment]
Thanks for the feedback!

> Lots of the emoji are not working.

That was 💄.🤘. Unnecessary so I've just removed them.

> Do you want to wrap references to `glamorous` in code tags? Not sure if that'll become obnoxious after a while. 😅

Yeah... Probably would become obnoxious...

> 🙌. I don't know if you abbreviated my name for some reason, but you're welcome to use my whole name if you like.😄

I actually copy pasted this from a document I made while on the plane. I meant to update it with your full name. I just couldn't remember how to spell it when on the plane 😅.

All your other comments are appreciated and have been applied. Thanks!

**Comment by kentcdodds** [comment]
I've updated the draft with more stuff. Now I've gotta [drive](https://github.com/kentcdodds/ama/issues/295). Wish me luck!

**Comment by paulmolluzzo** [comment]
>  I just couldn't remember how to spell it when on the plane 😅.

It's OK. I forget how to spell it sometimes too. 😂

**Comment by paulmolluzzo** [comment]
In the "Theme arg removed" section there's a code example that might better illustrate the value with a small update:

```js
const MyDynamicallyStyledSpan = glamorous
  .span(({ theme }) => ({
    fontSize: 20,
    color: theme.colors.primary
  }))
```

**Comment by luke-john** [comment]
Would it be possible to get another rc released?

I'm presenting glamorous at campjs this weekend and as they have a limited internet connection they have asked speakers to provide starter repos and global modules that attendees can download before hand to take the load off their limited internet connection.

Thanks.

**Comment by kentcdodds** [comment]
```
~/Developer/glamorous (next)
🐑  $ npm publish --tag beta
+ glamorous@4.0.0-beta.1
```

**Comment by paulmolluzzo** [comment]
Would it be helpful to provide some possible copy for the blog post?

Just in case you agree to it, here's some content for `shouldClassNameUpdate` (It's mostly from the README with a few changes to make it clearer I think):

## shouldClassNameUpdate

Most of the time, `glamor` is super fast, but scenarios where the styles are not dynamic it may be nice to prevent `glamor` from computing your styles on each render (and rerender). In these cases, you can implement `shouldClassNameUpdate` and if the component is rerendered the class name is not recalculated by `glamor`. Instead the same class name is reused.

For example:

```js
const pureDivFactory = glamorous('div', {
  shouldClassNameUpdate(props, previousProps, context, previousContext) {
    // return `true` to update the classname and
    // `false` to skip updating the class name
    return true
  },
})
const Div = pureDivFactory({marginLeft: 1})
render(<Div css={{marginLeft: 2}} />)
// this will render:
// <div />
// with {marginLeft: 2}
```

Note that this is not the same as `shouldComponentUpdate` and your component will still be rerendered at the appropriate times. `shouldClassNameUpdate` simply provides a means to opt-out of re-generating the class name unnecessarily.

**Comment by kentcdodds** [comment]
YES! Helping with the blog post is VERY welcome. I'm so tied up with so many things that it's taking me forever to make this happen. Sorry about that!

**Comment by paulmolluzzo** [comment]
👍 I'll try to post something for the other TODO sections here to copy/paste.

**Comment by paulmolluzzo** [comment]
Here's another section. I took a bit of editorial liberty here since the README at glamorous-primitives was a little light. If @nitin42 wants to take a look that'd be great!

---

## glamorous-primitives

`glamorous-primitives` is an independent library from [Nitin Tulswani](https://github.com/nitin42) that combines `glamorous` with [`react-primitives`](https://github.com/lelandrichardson/react-primitives) to provide a means of writing components that are usable across multiple target platforms.

Here is an example from the README:

```js
import React from 'react'
import { StyleSheet } from 'react-primitives'
import glamorous from 'glamorous-primitives'

const StyledView = glamorous.view({
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  marginTop: '80px'
})

const StyledText = glamorous.text({
  color: 'red',
  backgroundColor: 'mistyrose',
  fontSize: '40px'
})

export default class App extends React.Component {
  render () {
    return (
      <StyledView>
        <StyledText>Hello World!!</StyledText>
      </StyledView>
    )
  }
}
```

The library is a WIP and has plans to include a native constructor enabling rendering into Sketch. This is just one of the [related projects](https://github.com/paypal/glamorous#related-projects) the `glamorous` community has built.

**Comment by paulmolluzzo** [comment]
Here's another section write up for `glamorous-native`. I'd love if @ajwhite could review if he has a chance. 😄 

---

## [`glamorous-native`](https://github.com/robinpowered/glamorous-native)

`glamorous-native` is a port of `glamorous` that specifically targets React Native, exporting the same core functionality as `glamorous` as well as built-in component factories for all React Native elements. For every React Native element, there is an associated `glamorous` component factory attached to the `glamorous` function. You can access these factories like so: `glamorous.view`, `glamorous.text`, `glamorous.listView`, etc.

```js
const MyStyledView = glamorous.view({margin: 1})

<MyStyledView>{content}</MyStyledView>
// rendered output: <View style={{margin: 1}}>{content}</View>
// styles applied: {margin: 1}
// style is also cached and registered with `StyleSheet`
```
This project is actively maintained and has several [examples](https://github.com/robinpowered/glamorous-native/tree/master/examples) to learn from if you're interested in trying `glamorous-native` or new to native development with React Native entirely.

**Comment by paulmolluzzo** [comment]
Here's another section:

---

## Readable Class Names

With the release of v2.20.32, `glamor` introduced the ability to create readable class names for components by defining a `label` in the style object. The result is a generated class name with the pattern `css-[label]-[hash]`.

As an example:

```js
// App
const myRedText = glamorous.div({
  label: 'labelForRedText',
  color: 'red',
})

function App() {
  return (
    <myRedText>Whoa!</myRedText>
  )
}

render(<App />, document.getElementById('root'))

// result

<div tabindex="0" class="css-labelforredtext-1cz491u">Whoa!</div>
```

**Comment by paulmolluzzo** [comment]
@kentcdodds I'm not sure what you have in mind for Performance Optimizations. (it's relative to what you've already written about I would think.)

Do you want to talk about `glamorous/tiny`? (https://github.com/paypal/glamorous/pull/69)

What about Optimize getGlamorClassName (https://github.com/paypal/glamorous/pull/254)?

There's also this PR: https://github.com/paypal/glamorous/pull/64, but frankly I don't know enough about how `Object.freeze` and using string concatenation for the class names are improving performance.

**Comment by luke-john** [comment]
Also looks like the typescript screen capture is from an older version.

If you could update to the following that would be great thanks

again2

**Comment by nitin42** [comment]
## glamorous-primitives

**Demo** - 

demo

**Code**

**Comment by nitin42** [comment]
Link to the project - https://github.com/nitin42/glamorous-primitives

**Comment by kentcdodds** [comment]
I'm planning on releasing this tomorrow! So I'm going to work on getting everything ready! If you have any last minute items, please let me know!

**Comment by kentcdodds** [comment]
@paulmolluzzo, thanks so much for the content for the blog post!

**Comment by kentcdodds** [comment]
Ok! I've updated it! Would love reviews! https://medium.com/@kentcdodds/announcing-the-new-glamorous-website-and-v4-release-c678fe02a39a

**Comment by kentcdodds** [comment]
I'm thinking about merging this now and publishing the blog post tomorrow. That way if something goes wildly wrong we can address it. Thoughts?

**Comment by paulmolluzzo** [comment]
Under readable class names this can be improved:

> With the release of glamor v2.20.32, glamor introduced the ability to create readable class names for components by defining a label in the style object. 

Changed to:

> With the release of `v2.20.32`, `glamor` introduced the ability to create readable class names for components by defining a `label` in the style object.

Otherwise this looks great! 🎉

**Comment by kentcdodds** [comment]
AHHHHHHHHHHHHHHHHHHHHHHHHHHHHH! I JUST MERGED IT!!!!!!!!!!!!!!!!!


WAHOOOOO


In case it's not clear, that wasn't on accident. I'm just excited :tada:

**Comment by kentcdodds** [comment]
Release notes updated: https://github.com/paypal/glamorous/releases/tag/v4.0.0

Will publish the blog post tomorrow morning!
